File: Java/src/main/java/com/DSA/Java/Graphs/DisjointSet/DisJointSetImpl.java
package com.DSA.Java.Graphs.DisjointSet;
import java.util.ArrayList;

// Test Cases
// DisJointSetImpl newSet = new DisJointSetImpl();

// ArrayList<WeightedNode> nodeList = new ArrayList<>();
// nodeList.add(new WeightedNode("A"));
// nodeList.add(new WeightedNode("B"));
// nodeList.add(new WeightedNode("C"));
// nodeList.add(new WeightedNode("D"));

// newSet.makeSet(nodeList);
// WeightedNode firstNode = nodeList.get(0);
// WeightedNode secondNode = nodeList.get(1);
// DisJointSetImpl output = newSet.findSet(secondNode);
// output.printAllNodedOfSet();

// output = newSet.union(firstNode, secondNode);
// output.printAllNodedOfSet();

public class DisJointSetImpl {
  private ArrayList<WeightedNode> nodeList;

  public DisJointSetImpl() {
    this.nodeList = new ArrayList<>();
  }

  public void makeSet(ArrayList<WeightedNode> nodeList) {
    for (WeightedNode node : nodeList) {
      DisJointSetImpl set = new DisJointSetImpl();
      set.nodeList.add(node);
      node.set = set;
    }
  }

  public DisJointSetImpl findSet(WeightedNode node) {
    return node.set;
  }

  public DisJointSetImpl union(WeightedNode node1, WeightedNode node2) {
    if (node1.set.equals(node2.set)) {
      return null;
    } else {
      DisJointSetImpl set1 = node1.set;
      DisJointSetImpl set2 = node2.set;

      if (set1.nodeList.size() > set2.nodeList.size()) {
        ArrayList<WeightedNode> nodeSet2 = set2.nodeList;
        for (WeightedNode node : nodeSet2) {
          node.set = set1;
          set1.nodeList.add(node);
        }
        return set1;
      } else {
        ArrayList<WeightedNode> nodeSet1 = set1.nodeList;
        for (WeightedNode node : nodeSet1) {
          node.set = set2;
          set2.nodeList.add(node);
        }
        return set2;
      }
    }
  }

  public void printAllNodedOfSet() {
    System.out.print("Printing all nodes of the set: ");
    for (WeightedNode node : nodeList) {
      System.out.print(node.name + " ");
    }
    System.out.println();
  }
}

File: C++/Graph/src/DisjointSet/DisjointSetImpl.cpp
#include "../../include/DisjointSet/DisjointSetImpl.h"
#include "../../include/DisjointSet/WeightedNodeForSet.h"

void DisjointSetImpl::make_set(std::vector<WeightedNodeForSet *> &node_list) {
  for (const auto &node : node_list) {
    DisjointSetImpl *set = new DisjointSetImpl();
    set->node_list.push_back(node);
    node->set = set;
  }
}

DisjointSetImpl *DisjointSetImpl::find_set(WeightedNodeForSet *node) const {
  return node->set;
}

DisjointSetImpl *DisjointSetImpl::union_of_set(WeightedNodeForSet *node_1, WeightedNodeForSet *node_2) {
  if (node_1 == node_2) return nullptr;
  DisjointSetImpl *set_1 = node_1->set;
  DisjointSetImpl *set_2 = node_2->set;

  if (set_1->node_list.size() > set_2->node_list.size()) {
    std::vector<WeightedNodeForSet *> node_set_2 = set_2->node_list;
    for (const auto &node : node_set_2) {
      node->set = set_1;
      set_1->node_list.push_back(node);
    }
    return set_1;
  } else {
    std::vector<WeightedNodeForSet *> node_set_1 = set_1->node_list;
    for (const auto &node : node_set_1) {
      node->set = set_2;
      set_2->node_list.push_back(node);
    }
    return set_2;
  }
}

void DisjointSetImpl::print_all_nodes_of_set() {
  std::cout << "Printing all nodes of the set: " << std::endl;
  for (const auto &node : node_list) {
    std::cout << node->name << " ";
  }
  std::cout << "\n";
}
